Sources/X509/CryptographicMessageSyntax/CMSOperations.swift

```diff
@@ -0,0 +1,240 @@
+            // This force-unwrap is safe: we know there are certificates because we've located at least one certificate from this set!
+            var untrustedIntermediates = CertificateStore(signedData.certificates!)
+            untrustedIntermediates.insert(contentsOf: additionalIntermediateCertificates)
+            
+            var verifier = Verifier(rootCertificates: trustRoots, policy: policy)
+            let result = await verifier.validate(leafCertificate: signingCert, intermediates: untrustedIntermediates)
+
+            switch result {
+            case .validCertificate:
+                return .validSignature(.init(signer: signingCert))
+            case .couldNotValidate(let validationFailures):
+                return .unableToValidateSigner(.init(validationFailures: validationFailures, signer: signingCert))
+            }
+        } catch {
+            return .invalidCMSBlock(.init(reason: String(describing: error)))
+        }
+    }
+
+    @_spi(CMS)
+    public enum Error: Swift.Error {
+        case incorrectCMSVersionUsed
+    }
+
+    @_spi(CMS)
+    public enum SignatureVerificationResult {
+        case validSignature(Valid)
+        case unableToValidateSigner(SignerValidationFailure)
+        case invalidCMSBlock(InvalidCMSBlock)
+
+        public struct Valid: Hashable {
+            public var signer: Certificate
+
+            @inlinable
+            public init(signer: Certificate) {
+                self.signer = signer
+            }
+        }
+
+        public struct SignerValidationFailure: Hashable, Swift.Error {
+            public var validationFailures: [VerificationResult.PolicyFailure]
+
+            public var signer: Certificate
+
+            @inlinable
+            public init(validationFailures: [VerificationResult.PolicyFailure], signer: Certificate) {
+                self.validationFailures = validationFailures
+                self.signer = signer
+            }
+        }
+
+        public struct InvalidCMSBlock: Hashable, Swift.Error {
+            public var reason: String
+
+            @inlinable
+            public init(reason: String){
+                self.reason = reason
+            }
+        }
+
+        @inlinable
+        internal init(invalidCMSBlockReason: String) {
+            self = .invalidCMSBlock(.init(reason: invalidCMSBlockReason))
+        }
+    }
+}
+
+extension Array where Element == Certificate {
+    @usableFromInline
+    func certificate(signerInfo: CMSSignerInfo) throws -> Certificate? {
+        switch signerInfo.signerIdentifier {
+        case .issuerAndSerialNumber(let issuerAndSerialNumber):
+            for cert in self {
+                if cert.issuer == issuerAndSerialNumber.issuer && cert.serialNumber == issuerAndSerialNumber.serialNumber {
+                    return cert
+                }
+            }
+        case .subjectKeyIdentifier:
+            // This is unsupported for now.
```

Comment:
Should we throw instead for now?

Reply:
We will, because we'll fail to find a signer.